test/test_model.py

```diff
@@ -931,29 +933,79 @@ def test_relation_tool_errors(self):
             ops.model.ModelError,
-            [['relation-get', '-r', '3', '-', 'remote/0', '--app=False', '--format=json']],
+            [['relation-get', '-r', '3', '-', 'remote/0', '--format=json']],
         ), (
             lambda: fake_script(self, 'relation-get', 'echo {} >&2 ; exit 2'.format(err_msg)),
             lambda: self.backend.relation_get(3, 'remote/0', is_app=False),
             ops.model.RelationNotFoundError,
-            [['relation-get', '-r', '3', '-', 'remote/0', '--app=False', '--format=json']],
+            [['relation-get', '-r', '3', '-', 'remote/0', '--format=json']],
+        ), (
+            lambda: None,
+            lambda: self.backend.relation_get(3, 'remote/0', is_app=True),
+            ops.model.RelationNotFoundError,
+            [['relation-get', '-r', '3', '-', 'remote/0', '--app', '--format=json']],
         )]
 
-        for do_fake, run, exception, calls in test_cases:
-            do_fake()
-            with self.assertRaises(exception):
-                run()
-            self.assertEqual(fake_script_calls(self, clear=True), calls)
+        for i, (do_fake, run, exception, calls) in enumerate(test_cases):
+            with self.subTest(i):
+                do_fake()
+                with self.assertRaises(exception):
+                    run()
+                self.assertEqual(fake_script_calls(self, clear=True), calls)
+
+    def test_relation_get_juju_version_quirks(self):
+        self.addCleanup(os.environ.pop, 'JUJU_VERSION', None)
+
+        fake_script(self, 'relation-get', '''echo '{"foo": "bar"}' ''')
+
+        # on 2.7.0+, things proceed as expected
+        for v in ['2.8.0', '2.7.0']:
+            with self.subTest(v):
+                os.environ['JUJU_VERSION'] = v
+                rel_data = self.backend.relation_get(1, 'foo/0', is_app=True)
+                self.assertEqual(rel_data, {"foo": "bar"})
+                calls = [' '.join(i) for i in fake_script_calls(self, clear=True)]
+                self.assertEqual(calls, ['relation-get -r 1 - foo/0 --app --format=json'])
+
+        # before 2.7.0, it just fails (no --app support)
+        os.environ['JUJU_VERSION'] = '2.6.9'
+        with self.assertRaisesRegex(RuntimeError, 'Juju too old'):
+            self.backend.relation_get(1, 'foo/0', is_app=True)
+        self.assertEqual(fake_script_calls(self), [])
+
+    def test_relation_set_juju_version_quirks(self):
+        self.addCleanup(os.environ.pop, 'JUJU_VERSION', None)
+
+        fake_script(self, 'relation-set', 'exit 0')
+
+        # on 2.7.0+, things proceed as expected
+        for v in ['2.8.0', '2.7.0']:
+            with self.subTest(v):
+                os.environ['JUJU_VERSION'] = v
+                self.backend.relation_set(1, 'foo', 'bar', is_app=True)
+                calls = [' '.join(i) for i in fake_script_calls(self, clear=True)]
+                self.assertEqual(calls, ['relation-set -r 1 foo=bar --app'])
```

Comment:
If we feel like this is cleaner, can we just have that be the formatted version of fake_script_calls ?
Or is that what you wanted to do in a separate patch? (I think I like it more, as it is easier to write and read the tests and we don't care that sometimes the string may not be unambiguously parsable.)

Reply:
it wasn't part of what I was looking to refactor, but it does make things more readable for me.
I'll keep it on the backburner :-)